pkg/noderesourcetopology/least_numa.go

```diff
@@ -0,0 +1,157 @@
+/*
+Copyright 2023 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package noderesourcetopology
+
+import (
+	"fmt"
+
+	v1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/api/resource"
+	"k8s.io/klog/v2"
+	v1qos "k8s.io/kubernetes/pkg/apis/core/v1/helper/qos"
+	"k8s.io/kubernetes/pkg/kubelet/cm/topologymanager/bitmask"
+	"k8s.io/kubernetes/pkg/scheduler/framework"
+
+	topologyv1alpha1 "github.com/k8stopologyawareschedwg/noderesourcetopology-api/pkg/apis/topology/v1alpha1"
+	"gonum.org/v1/gonum/stat/combin"
+	"sigs.k8s.io/scheduler-plugins/pkg/util"
+)
+
+var zeroQty resource.Quantity
+
+func init() {
+	zeroQty = resource.MustParse("0")
+}
+
+func leastNUMAContainerScopeScore(pod *v1.Pod, zones topologyv1alpha1.ZoneList) (int64, *framework.Status) {
+	nodes := createNUMANodeList(zones)
+	qos := v1qos.GetPodQOS(pod)
+
+	maxNUMANodesCount := 0
+	// the order how TopologyManager asks for hint is important so doing it in the same order
+	// https://github.com/kubernetes/kubernetes/blob/master/pkg/kubelet/cm/topologymanager/scope_container.go#L52
+	for _, container := range append(pod.Spec.InitContainers, pod.Spec.Containers...) {
+		identifier := fmt.Sprintf("%s/%s/%s", pod.Namespace, pod.Name, container.Name)
+		numaNodes, err := numaNodesRequired(identifier, qos, nodes, container.Resources.Requests)
+		if err != nil {
+			return 0, framework.NewStatus(framework.Error, err.Error())
+		}
+
+		if numaNodes.Count() > maxNUMANodesCount {
+			maxNUMANodesCount = numaNodes.Count()
+		}
+
+		// subtract the resources requested by the container from the given NUMA.
+		// this is necessary, so we won't allocate the same resources for the upcoming containers
+		subtractFromNUMAs(container.Resources.Requests, nodes, numaNodes.GetBits()...)
+	}
+
+	return normalizeScore(maxNUMANodesCount), nil
+}
+
+func leastNUMAPodScopeScore(pod *v1.Pod, zones topologyv1alpha1.ZoneList) (int64, *framework.Status) {
+	nodes := createNUMANodeList(zones)
+	qos := v1qos.GetPodQOS(pod)
+
+	identifier := fmt.Sprintf("%s/%s", pod.Namespace, pod.Name)
+
+	resources := util.GetPodEffectiveRequest(pod)
+
+	numaNodes, err := numaNodesRequired(identifier, qos, nodes, resources)
```

Comment:
in reality, how possible `numaNodesRequired()` would return an error? and is this error _critical_?

I asked this is b/c in the scheduler framework, a score plugin's [failure](https://github.com/kubernetes/kubernetes/blob/9787f46bbf05349ca53cd8b04cd2f7a76771646a/pkg/scheduler/framework/runtime/framework.go#L939-L957) would cause [the whole scheduling](https://github.com/kubernetes/kubernetes/blob/9787f46bbf05349ca53cd8b04cd2f7a76771646a/pkg/scheduler/schedule_one.go#L371-L374) to entirely fail. So it's pretty rare that a Score plugin to return critical error.

Reply:
The only case when `numaNodesRequired` returns error is when the container/pod requires more resources than available on the worker node, so it's rather impossible since score plugin should  run after `NodeResourcesFit` plugin. Instead of an error I can print warning message and return `minScore`(bitmask with all 8 bits set) here.

Reply:
So I have changed this function signature to return only `bitmask.Bitmask`. The problem I have is that right now I have chosen to return `nil` when we can't fit resources onto the worker node and just check for `nil` to return quickly(in container scope it's necessary not to attempt subtractFromNUMAs):
```go
		if numaNodes == nil {
			return framework.MinNodeScore, nil
		}
```
I am not fully satisfied with this approach but I can't think of any simpler solution either 😞 
I was playing around with filling bitmask with all bits up to `highestNUMAiD` but it is very clumsy:
```go
allNodesBitmask, _ := bitmask.NewBitMask(0, 1, 2, 3, 4, 5, 6, 7, 8)
```

Reply:
> so it's rather impossible since score plugin should run after NodeResourcesFit plugin

This is very true. Maybe add a comment here:

```go
if numaNodes == nil {
	// Should not reach here.
	return framework.MinNodeScore, nil
}
```